sycl/unittests/accessor/AccessorIterator.cpp

```diff
@@ -0,0 +1,400 @@
+      ASSERT_EQ(copied[i], reference[i]);
+    }
+  }
+
+  template <typename T, typename IteratorT>
+  std::vector<T> copyThroughIterators(IteratorT begin, IteratorT end) {
+    std::vector<T> copied;
+    for (auto it = begin; it != end; ++it)
+      copied.push_back(*it);
+
+    return copied;
+  }
+
+  template <int Dimensions, typename T = int>
+  void
+  validatePartialCopyThroughIterator(const std::vector<T> &copied,
+                                     sycl::buffer<T, Dimensions> &buffer,
+                                     const sycl::range<Dimensions> &copyShape,
+                                     const sycl::id<Dimensions> &offset = {}) {
+    auto fullAccessor = buffer.template get_access<sycl::access_mode::read>();
+    size_t linearId = 0;
+
+    sycl::id<3> offsetToUse(Dimensions > 2 ? offset[Dimensions - 3] : 1,
+                            Dimensions > 1 ? offset[Dimensions - 2] : 1,
+                            offset[Dimensions - 1]);
+
+    sycl::id<3> shapeToCheck(
+        (Dimensions > 2 ? copyShape[Dimensions - 3] : 1) + offsetToUse[0],
+        (Dimensions > 1 ? copyShape[Dimensions - 2] : 1) + offsetToUse[1],
+        copyShape[Dimensions - 1] + offsetToUse[2]);
+
+    for (size_t z = offsetToUse[0]; z < shapeToCheck[0]; ++z) {
+      for (size_t y = offsetToUse[1]; y < shapeToCheck[1]; ++y) {
+        for (size_t x = offsetToUse[2]; x < shapeToCheck[2]; ++x) {
+          auto value = accessHelper<Dimensions>(fullAccessor, z, y, x);
+          ASSERT_EQ(copied[linearId], value);
+          ++linearId;
+        }
+      }
+    }
+  }
+
+  template <int TotalDimensions, int CurrentDimension = 3, typename Container,
+            typename... Indices>
+  auto &&accessHelper(Container &&C, int Idx, Indices... Ids) {
+    if constexpr (CurrentDimension > TotalDimensions) {
+      (void)Idx;
+      return accessHelper<TotalDimensions, CurrentDimension - 1>(C, Ids...);
+    } else
+      return accessHelper<TotalDimensions, CurrentDimension - 1>(C[Idx],
+                                                                 Ids...);
+  }
+
+  template <int TotalDimensions, int CurrentDimension = 3, typename Container>
+  auto &&accessHelper(Container &&C, int Idx) {
+    return C[Idx];
+  }
+};
+
+TEST_F(AccessorIteratorTest, ImplementationDetails) {
+  std::vector<int> reference(5);
+  std::iota(reference.begin(), reference.end(), 0);
+  sycl::buffer<int> buffer(reference.data(), sycl::range<1>{reference.size()});
+  auto accessor = buffer.template get_access<sycl::access_mode::read_write>();
+  {
+    auto It = accessor.begin();
+    // Check that It can't be decremented past begin
+    ASSERT_EQ(--It, accessor.begin());
+    ASSERT_EQ(It - 1, accessor.begin());
+    ASSERT_EQ(It -= 1, accessor.begin());
+    ASSERT_EQ(It - 10, accessor.begin());
+    ASSERT_EQ(It -= 10, accessor.begin());
+  }
+  {
+    auto It = accessor.end();
+    // Check that It can't be incremented past end
```

Comment:
Is it in the SYCL specification?

Reply:
It is neither a part of SYCL specification nor it is a part of C++ specification. This is a test for an implementation detail. Essentially, C++ says that decrementing past `.begin()` or incrementing past `.end()` is a UB. By some reason, I decided to make this UB more user-friendly.

I agree that it is not strictly necessary and it makes sense to remove associated checks from the implementation.

Reply:
Removed the test and related parts of the implementation in 8e2db4ef3f3a6c95c161a5b4d0825ce1a3173fa0

Reply:
I understand your motivation.
But UB in C/C++ are not there to annoy the user but to allow bare-metal speed when needed. At the end the goal is not to have UB anyway and less speed. :-)
Safe users or coding rules can avoid the use of pointers and iterators and use `std::span`, `std::ranges`, etc.